blocks/search.py

```diff
@@ -0,0 +1,261 @@
+from abc import ABCMeta, abstractmethod
+from collections import OrderedDict
+
+import numpy as np
+
+from theano import config
+from theano import function
+from theano import tensor
+
+from blocks.filter import VariableFilter
+from blocks.graph import ComputationGraph
+
+floatX = config.floatX
+
+
+class Search(object):
+    """Abstract search class
+
+    Parameters
+    ----------
+    sequence_generator : sequence generator
+        Sequence generator to be used
+
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, sequence_generator):
+        self.generator = sequence_generator
+        self.compiled = False
+
+    @abstractmethod
+    def compile(self):
+        self.compiled = True
+
+    @abstractmethod
+    def search(self, beam_size, **kwargs):
+        r"""Performs search
+
+        Parameters
+        ----------
+        \*\*kwargs :
+            Arguments needed by sequence generator
+
+        Returns
+        -------
+        Generated sequences
+        """
+        if not self.compiled:
+            self.compile()
+
+
+class GreedySearch(Search):
+
+    def __init__(self, sequence_generator):
+        super(GreedySearch, self).__init__(sequence_generator)
+
+
+class BeamSearch(Search):
+    """Beam search
+
+    Parameters
+    ----------
+    beam_size : int
+        Size of beam
+    batch_size : int
+        Size of input batch
+    sequence_generator : sequence generator
+        Sequence generator brick
+
+    """
+    def __init__(self, beam_size, batch_size, sequence_generator, attended,
+                 y, attended_mask, y_mask, inputs_dict, inputs):
```

Comment:
Yes, thougth about it, but the current way seemed easier to me. I'll rewrite this part.